netket/sampler/autoreg.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from flax import struct
+from netket.sampler import Sampler, SamplerState
+from netket.utils.types import PRNGKeyT, PyTree
+
+import jax
+from jax import numpy as jnp
+
+
+def vmap_choice(key, a, p, replace=True):
+    """
+    p.shape: (batch, a.shape)
+    Return shape: (batch, )
+    """
+    def scan_fun(key, p_i):
+        new_key, key = jax.random.split(key)
+        out_i = jax.random.choice(key, a, replace=replace, p=p_i)
+        return new_key, out_i
+
+    _, out = jax.lax.scan(scan_fun, key, p)
+    return out
+
+
+@struct.dataclass
+class ARSamplerState(SamplerState):
+    model_state: PyTree
+    key: PRNGKeyT
+
+    def __repr__(self):
+        return f'ARSamplerState(rng state={self.key})'
+
+
+@struct.dataclass
+class ARSampler(Sampler):
+    """Sampler for autoregressive neural networks."""
+    def _init_state(sampler, model, params, key):
+        return ARSamplerState(model_state=None, key=key)
```

Comment:
here you should setup the initial state of the samples by calling `self.hilbert.random_state(key, self.n_chains)` or something